test/extensions/filters/network/rbac/integration_test.cc

```diff
@@ -133,6 +133,30 @@ name: rbac
   EXPECT_EQ(0U, test_server_->counter("tcp.rbac.shadow_denied")->value());
 }
 
+TEST_P(RoleBasedAccessControlNetworkFilterIntegrationTest, DelayDenied) {
+  initializeFilter(R"EOF(
+name: rbac
+typed_config:
+  "@type": type.googleapis.com/envoy.extensions.filters.network.rbac.v3.RBAC
+  stat_prefix: tcp.
+  rules:
+    policies:
+      "deny_all":
+        permissions:
+          - any: true
+        principals:
+          - not_id:
+              any: true
+  delay_deny: 0.5s
+)EOF");
+  IntegrationTcpClientPtr tcp_client = makeTcpConnection(lookupPort("listener_0"));
+  ASSERT_TRUE(tcp_client->write("hello", false, false));
+  tcp_client->waitForDisconnect();
```

Comment:
This does not really test that disconnect occurred after configured delay. You can use simulated time for that. Check out MultiplexedIntegrationTestWithSimulatedTime

Reply:
@yanavlasov  I checked `MultiplexedIntegrationTestWithSimulatedTime` and found it uses the `Event::TestUsingSimulatedTime` but not quite understand how it works.

The `MultiplexedIntegrationTestWithSimulatedTime` uses `timeSystem().advanceTimeWait` in its test case but the same statement doesn't seem working in the RBAC integration:

I subclassed `TestUsingSimulatedTime` and tried the same but it just never triggers the timer in the RBAC filter as the `tcp_client->waitForDisconnect()` is blocked forever, also tried `simTime().advanceTimeWait` but it is still not working.

would you mind expand a little bit how to use the simulated time in the integration test? thanks.

Reply:
It should work. Can you update PR with the code that uses simulated time, please? I can comment to address problems.

Note that the `partialWrite` function will not work with simulated time, since it has a timer. However I do not think you actually need it. Why not just use write?

If you want to check that no data was actually proxied, you can just check on the upstream server.

Reply:
@yanavlasov I added back the test code in `DelayDeniedSimulatedTimer`, it will timeout at line 157  (`tcp_client->waitForDisconnect();`) that seems like the simulated timer is never triggered. I'm not sure if I missed something, I checked the `MultiplexedIntegrationTestWithSimulatedTime` but didn't notice anything different regarding the use of the simulated timer, please let me know if you have any ideas, thanks.

As regarding the `partialWrite`, it is to test the `readDisable(true)` by checking the data is not fully read, it is suggested by @lambdai , I'm fine to remove it if you think it's not needed though (we still have the unit test to check we are calling `readDisable(true)`).

Reply:
This work for me on my workstation. I've opened #35128 to see what may be different.